pyface/font.py

```diff
@@ -0,0 +1,478 @@
+            # substitute synonyms
+            value = parser_synonyms.get(value, value)
+            if value.lower() in weights:
+                if weight != 'normal':
+                    raise FontParseError(
+                        f"Weight declared twice in {description!r}"
+                    )
+                weight = value.lower()
+            elif value.lower() in stretches:
+                if stretch != 'normal':
+                    raise FontParseError(
+                        f"Stretch declared twice in {description!r}"
+                    )
+                stretch = value.lower()
+            elif value.lower() in sizes:
+                if size != -1:
+                    raise FontParseError(
+                        f"Size declared twice in {description!r}"
+                    )
+                size = sizes[value.lower()]
+            elif value.lower() in styles:
+                if style != 'normal':
+                    raise FontParseError(
+                        f"Style declared twice in {description!r}"
+                    )
+                style = value.lower()
+            elif value in variants:
+                if value.lower() in variant_set:
+                    raise FontParseError(
+                        f"Variant {value!r} declared twice in {description!r}"
+                    )
+                variant_set.add(value.lower())
+            else:
+                # assume it is a font family name
+                family.append(value)
+        elif kind == 'QUOTED_NAME':
+            family.append(value[1:-1])
+        elif kind == 'MISMATCH':
+            raise FontParseError(
+                f"Parse error {value!r} at {index} in {description!r}"
+            )
+    if len(family) == 0:
+        family = ['default']
+    if size == -1:
+        size = 12.0
+    return {
+        'family': family,
+        'weight': weight,
+        'stretch': stretch,
+        'style': style,
+        'variants': variant_set,
+        'size': size,
+    }
+
+
+class Font(HasStrictTraits):
+    """ A toolkit-independent font specification. """
+
+    #: The preferred font families.
+    family = FontFamily()
+
+    #: The weight of the font.
+    weight = FontWeight()
+
+    #: How much the font is expanded or compressed.
+    stretch = FontStretch()
+
+    #: The style of the font.
+    style = FontStyle()
+
+    #: The size of the font.
+    size = FontSize()
+
+    #: The font variants.
+    variants = FontVariants()
+
+    @classmethod
+    def from_description(cls, description):
+        """ An extremely lax 'parser' for CSS-style font descriptions.
+
+        Parameters
+        ----------
+        description : str
+            A font description in string form such as
+            'italic bold 14pt Helvetica, Arial, sans-serif' or
+            '36pt "Comic Sans"'
+        """
+        return cls(**parse_font_description(description))
+
+    @classmethod
+    def from_toolkit(cls, toolkit_font):
```

Comment:
This is because there are going to be situations (like font dialogs, but not only there), where there is some toolkit-level call or object which returns a toolkit font object (either a `QFont` or a `wx.Font`).  We need to be able to convert those to a `Font` object as best we can.  It may not be precise, because different backends have different capabilities, but is should be good enough, and should round-trip within the capabilities of the system.

The fact that it isn't precise is OK because font selection is strongly constrained by what is actually available on on the underlying system, so even if the font selection is not exactly the same, you are likely to get the same typeface when the toolkit asks for things.

Reply:
That makes sense. Thank you! Resolving.